tests/pq/test_pgconn.py

```diff
@@ -389,6 +393,100 @@ def test_set_single_row_mode(pgconn):
     pgconn.set_single_row_mode()
 
 
+@contextlib.contextmanager
+def cancellable_query(pgconn: PGconn, monitor_conn: PGconn) -> Iterator[None]:
+    pgconn.send_query_params(b"SELECT pg_sleep($1)", [b"180"])
+    while True:
+        r = monitor_conn.exec_(
+            b"SELECT count(*) FROM pg_stat_activity"
+            b" WHERE query = 'SELECT pg_sleep($1)'"
+            b" AND state = 'active'"
+        )
+        assert r.status == pq.ExecStatus.TUPLES_OK
+        if r.get_value(0, 0) != b"0":
+            break
+
+        time.sleep(0.01)
+
+    yield None
+
+    res = pgconn.get_result()
+    assert res is not None
+    assert res.status == pq.ExecStatus.FATAL_ERROR
+    assert res.error_field(pq.DiagnosticField.SQLSTATE) == b"57014"
+    while pgconn.is_busy():
+        pgconn.consume_input()
+
+
+@pytest.mark.libpq(">= 17")
+def test_cancel_nonblocking(dsn):
+    # mimic test_cancel() from src/test/modules/libpq_pipeline/libpq_pipeline.c
+    def connect() -> PGconn:
+        conn = pq.PGconn.connect(dsn.encode())
+        if conn.status != pq.ConnStatus.OK:
+            pytest.fail(
+                f"bad connection: {conn.error_message.decode('utf8', 'replace')}"
+            )
+        return conn
+
+    conn, monitor_conn = connect(), connect()
+    conn.nonblocking = 1
+
+    # test PQcancel
+    with cancellable_query(conn, monitor_conn):
+        cancel = conn.get_cancel()
+        cancel.cancel()
+
+    # PGcancel object can be reused for the next query
+    with cancellable_query(conn, monitor_conn):
+        cancel.cancel()
+
+    del cancel
+
+    # test PQcancelBlocking
+    with cancellable_query(conn, monitor_conn):
+        cancel_conn = conn.cancel_conn()
+        assert cancel_conn.status == pq.ConnStatus.ALLOCATED
+        cancel_conn.blocking()
+        assert cancel_conn.status == pq.ConnStatus.OK  # type: ignore[comparison-overlap] # noqa: E501
+    cancel_conn.finish()
+    del cancel_conn
+
+    def wait(cancel_conn: PGcancelConn) -> None:
```

Comment:
Don't we have something similar to this already in the test suite that we can refactor/reuse?

Reply:
Yes we do, in this very file, in `test_connect_async()`, `test_connect_async_bad()` and `test_reset_async()` ! I'll try to factor out a common helper in a follow-up.

Reply:
Done in a preparatory commit 005e9b1bc593fc003b8cdb241b323bc5d55d85be